File: typescript/tutorials/tour-of-workflows-typescript/src/workflows/signup-with-timers.ts
import * as restate from "@restatedev/restate-sdk";
import {
  RestatePromise,
  TerminalError,
  WorkflowContext,
  WorkflowSharedContext,
} from "@restatedev/restate-sdk";
import { sendVerificationEmail, sendReminderEmail, User } from "../utils";

export const signupWithTimers = restate.workflow({
  name: "SignupWithTimersWorkflow",
  handlers: {
    run: async (ctx: WorkflowContext, user: User) => {
      const userId = ctx.key;

      const secret = ctx.rand.uuidv4();
      await ctx.run("verify", () =>
        sendVerificationEmail(userId, user, secret),
      );

      const clickedPromise = ctx.promise<string>("email-verified").get();
      const verificationTimeout = ctx.sleep({ days: 1 });
      while (true) {
        const reminderTimer = ctx.sleep({ seconds: 15 });

        // Wait for email verification, reminder timer or timeout
        const result = await RestatePromise.race([
          clickedPromise.map(() => "verified"),
          reminderTimer.map(() => "reminder"),
          verificationTimeout.map(() => "timeout"),
        ]);

        switch (result) {
          case "verified":
            const clickedSecret = await clickedPromise;
            return { success: clickedSecret === secret };
          case "reminder":
            await ctx.run("send reminder", () =>
              sendReminderEmail(userId, user, secret),
            );
            break;
          case "timeout":
            throw new TerminalError(
              "Email verification timed out after 24 hours",
            );
        }
      }
    },
    verifyEmail: async (
      ctx: WorkflowSharedContext,
      req: { secret: string },
    ) => {
      await ctx.promise<string>("email-verified").resolve(req.secret);
    },
  },
});

File: python/tutorials/tour-of-workflows-python/app/workflows/signup_with_timers.py
import restate
from restate import WorkflowContext, WorkflowSharedContext, TerminalError
from datetime import timedelta
from app.utils import (
    User,
    send_verification_email,
    send_reminder_email,
    VerifyEmailRequest,
)

signup_with_timers = restate.Workflow("SignupWithTimersWorkflow")


@signup_with_timers.main()
async def run(ctx: WorkflowContext, user: User) -> bool:
    user_id = ctx.key()

    secret = str(ctx.uuid())
    await ctx.run_typed(
        "verify",
        send_verification_email,
        user_id=user_id,
        user=user,
        verification_secret=secret,
    )

    clicked_promise = ctx.promise("email-verified")
    verification_timeout = ctx.sleep(timedelta(days=1))

    while True:
        reminder_timer = ctx.sleep(timedelta(seconds=15))

        # Wait for email verification, reminder timer or timeout
        result = await restate.select(
            verification=clicked_promise.value(),
            reminder=reminder_timer,
            timeout=verification_timeout,
        )

        match result:
            case ["verification", clicked_secret]:
                return clicked_secret == secret
            case ["reminder", _]:
                await ctx.run_typed(
                    "remind",
                    send_reminder_email,
                    user_id=user_id,
                    user=user,
                    verification_secret=secret,
                )
            case ["timeout", _]:
                raise TerminalError("Email verification timed out after 24 hours")


@signup_with_timers.handler("verifyEmail")
async def verify_email(ctx: WorkflowSharedContext, req: VerifyEmailRequest) -> None:
    await ctx.promise("email-verified").resolve(req.secret)
